Add LegacyServiceAccountTokenCleanUp feature in beta (#43563)
kubernetes/website: content/en/docs/reference/access-authn-authz/service-accounts-admin.md

to obtain short-lived API access tokens is recommended instead.
{{< /note >}}

## Auto-generated legacy ServiceAccount token clean up

Starting with v1.29, auto-generated legacy ServiceAccount tokens will be marked
as invalid if they have not been used for a specified duration
(defaults to one year). If the invalidated tokens have not been used for a
specified duration (defaults to one year), they will be cleaned up by the

> control plane.

Reviewer: How do I parse these ?
It takes one year for the token to be marked invalid.
It then takes another year for the token to be cleaned up.
Right?

Author: Yes. I revised again to make it more clear.

1. If the spec of the incoming Pod doesn't already contain any `imagePullSecrets`, then the
   admission controller adds `imagePullSecrets`, copying them from the `ServiceAccount`.

### Legacy ServiceAccount token cleaner

{{< feature-state for_k8s_version="v1.29" state="beta" >}}

> Starting with Kubernetes version 1.28, a new feature called

Reviewer 2: This feature is already GA'ed. Just mention that in a different section:

### Legacy ServiceAccount token tracking controller

This controller generates a ConfigMap `kube-system/kube-apiserver-legacy-service-account-token-tracking` to record the timestamp when legacy service account tokens began to be monitored by the system.

Author: Done.

`LegacyServiceAccountTokenTracking` is enabled by default. This feature creates
a ConfigMap called `kube-apiserver-legacy-service-account-token-tracking` in the
`kube-system` namespace. This ConfigMap stores the current date as the tracking
date. With this feature enabled, each time a legacy service account token is
used, the Secret label `kubernetes.io/legacy-token-last-used` is updated with
the latest usage date.

Starting with Kubernetes version 1.29, the legacy ServiceAccount token cleaner
is enabled by default. It runs as part of the `kube-controller-manager` and
checks every 24 hours to see if any legacy service account token has not been
used in a *specified amount of time*. If so, the cleaner marks those tokens as
invalid.

The cleaner works by first checking the ConfigMap created by the

> `LegacyServiceAccountTokenTracking` feature. If the current time is a

Reviewer 3: ```suggestion
control plane (provided that `LegacyServiceAccountTokenTracking` is enabled).
If the current time is a
```

Author: Revised.

*specified amount of time* after the date in the ConfigMap, the cleaner then
loops through the list of Secrets in the cluster and evaluates each Secret that
has the type `kubernetes.io/service-account-token`.

If a Secret meets all of the following conditions, the cleaner marks it as
invalid:

- The Secret is auto-generated, meaning that it is bi-directionally referenced
  by a ServiceAccount.
- The Secret is not currently mounted by any pods.
- The Secret has not been used in a *specified amount of time* since it was
  created or since it was last used.

> The cleaner marks an invalid Secret by adding a label called

Reviewer: ```suggestion
The cleaner marks a Secret invalid by adding a label called
```

Author: Revised.

`kubernetes.io/legacy-token-invalid-since` to the Secret, with the current date
as the value. If an invalid Secret is not used in a *specified amount of time*,
the cleaner will delete it.

{{< note >}}
All the *specified amount of time* above is default to be one year. Users can
configure this value through the `--legacy-service-account-token-clean-up-period`

> flag.

Reviewer: on the kube-controller-manager component, I suppose?

Reviewer 3: ```suggestion
command line argument.
```
Outside of the Go ecosystem, “flag” refers to a boolean value (and this isn't a boolean).

Author: Revised.

{{< /note >}}

> If users use the invalidated auto-generated legacy token, the legacy token

Reviewer: We can avoid wordy phrases when there is not ambiguity.

```suggestion
If users use an invalidated token, the token
```

Author: Revised.

validator will

> - add an audit annotation for the key-value pair

Reviewer: for which "key-value pair"?

>   `authentication.k8s.io/legacy-token-invalidated`:`secretName/Namepace`,

Reviewer: ```suggestion
  `authentication.k8s.io/legacy-token-invalidated: <secret name>/<namepace>`,
```

Author: Revised.

- increment the `invalid_legacy_auto_token_uses_total` metric count,
- update the Secret label `kubernetes.io/legacy-token-last-used` with the new
  date,
- return an error indicating that the legacy token has been invalidated.

> If received this validation error, users can update the Secret to remove the

Reviewer: ```suggestion
When receiving this validation error, users can update the Secret to remove the
```

Author: Revised.

`kubernetes.io/legacy-token-invalid-since` label to temporarily allow use of
this token.

Here's an example of an auto-generated legacy token that has been marked with the
`kubernetes.io/legacy-token-last-used` and `kubernetes.io/legacy-token-invalid-since`
labels:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: build-robot-secret
  namespace: default
  labels:
    kubernetes.io/legacy-token-last-used: 2022-10-24
    kubernetes.io/legacy-token-invalid-since: 2023-10-25
  annotations:
    kubernetes.io/service-account.name: build-robot
type: kubernetes.io/service-account-token
```

{{< note >}}

> The difference between the manually created legacy token and
> the auto-generated legacy token is that only auto-generated legacy token is
> bi-directionally referenced by the ServiceAccount, like:

Reviewer 3: ```suggestion
The way that Kubernetes is able to tell the difference between a manually created
legacy token and an auto-generated legacy token is by assuming that only auto-generated
legacy tokens are bi-directionally referenced by the ServiceAccount. For example:
```

Author: Revised.

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: build-robot
  namespace: default

> secrets:
>   - name: build-robot-secret

Reviewer 3: ```suggestion
secrets:
  - name: build-robot-secret # triggers automatic legacy token generation
                             # (only on older clusters - legacy token generation is disabled by default)
                             # usually NOT present for a manually generated token
```

Reviewer 3: How about that?

Author: Revised.

```

> 

Reviewer 2: refer to the section "Create additional API tokens"

Author: Moved the note to the section "Create additional API tokens".

If you want to use the manually created legacy token, please do not add the

> Secret name to the `secrets` field of the service account.

Reviewer: ```suggestion
Secret name to the `secrets` field of the ServiceAccount.
```

Author: Revised.

1. If the spec of the incoming Pod doesn't already contain any `imagePullSecrets`, then the
   admission controller adds `imagePullSecrets`, copying them from the `ServiceAccount`.

### Legacy ServiceAccount token cleaner

{{< feature-state for_k8s_version="v1.29" state="beta" >}}

Starting from Kubernetes version 1.28, the feature `LegacyServiceAccountTokenTracking`
is enabled by default, which creates a ConfigMap that records the configmap

> creation date as the tracking date. With the feature enabled, each use of the

Reviewer: This is confusing. Why are we creating a ConfigMap in this context?
What's the the namespace, name of the ConfigMap?
We are simply creating a ConfigMap for recording a tracking date?

Author: Hi Qiming, thank you very much for your comments. I added some detailed information of the ConfigMap. I need to explain the ConfigMap here as it will be used by the cleaner to check the tracking date to determine whether it's time to start to evaluate the secrets. If I don't explain the ConfigMap, the user will be confused and wonder why the cleaner has not started to mark some secret as invalid.

Reviewer: This is an awful design, I'd say. ConfigMaps are containers for things that are meant to be exposed to users. Having a ConfigMap in the system, without knowing that it is being used by a builtin controller, I'd call this a confusing design. I'd prefer to record the timestamps as annotations on those objects to be cleaned. I believe many people like me would clean ConfigMaps we don't recognize.
If the design cannot be changed, we have to warn users that it is a special ConfigMap, don't touch it, don't modify it or remove it. I don't think really care about the facilities used for tracking, be it a ConfigMap or something else. All they may want to know is that there is a controller to track this.

Reviewer 2: the configmap is namespaced in `kube-system`. all stuff within `kube-system` is not supposed be modified/deleted if not knowing what's it is doing. e.g. do you want to delete `kube-system/kube-root-ca.crt`?

Reviewer: Does this ConfigMap have a well-known name?
I can guess the why the `kube-root-ca.crt` ConfigMap is there, and I won't touch it unless necessary.

Author: Hi Qiming, it's name is `kube-apiserver-legacy-service-account-token-tracking`. I have updated this name in the doc:)

Reviewer: sigh ...

legacy secret token updates the secret label `kubernetes.io/legacy-token-last-used`,
with the latest usage date as the value.

Starting from Kubernetes version 1.29, the legacy ServiceAccount token cleaner
is enabled by default. It runs as part of `kube-controller-manager`
and acts asynchronously. Every 24 hours, the cleaner checks if the current time

> has been a `speficied time` since the ConfigMap tracking date. If true, it loops

Reviewer: Please don't use backtiqs if you are referring something that is neither a component name, nor a command name or field name. Use italics would be appropriate.

Author: Revised.

> the secret list and evaluates the secrets that have the type `kubernetes.io/service-account-token`.

Reviewer: You mean "it identifies Secrets about ServiceAccount tokens" ...
I'd suggest we focus on the things users need to know. It is okay to explain how things work, but still, please focus on the implications to users.

For example, if the ConfigMap is kind of implementation details users don't need
to know, we can skip that. If the users may need to know it because something
bad would happen and they can mitigate those situations by manipulating this
ConfigMap, we can elaborate on that.

Author: Thank you, Qiming, explained above. Please refer this doc for the whole design if you are interested: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2799-reduction-of-secret-based-service-account-token#design-details

If the secret satisfies all the following conditions:
- is auto-generated (secret bi-directionally referenced by the ServiceAccount)
- is not mounted by pods
- is not used in a `specified time` since the last-used/secret-creation date

the cleaner marks the secret as invalid by adding a label `kubernetes.io/legacy-token-invalid-since`,
with the current date as value. If the invalidated token is not used in a

> `specified time`, it will be cleaned by the cleaner.

Reviewer: Are we talking about a LRU algorithm here?

Author: No, it not a LRU algorithm here. For a legacy token, if it is not used for more than one year, the cleaner marks it as invalid. After it is invalidated, if it is not used for another one year, the cleaner delete it. Marking the Secret invalid is a guard step in case the cleaner deletes some legacy tokens that are still wanted by the users.

{{< note >}}
All the `specificied time` above is default to be one year. Users can configure
this value through `--legacy-service-account-token-clean-up-period` flag.
{{< /note >}}

If users use the invalidated auto-generated legacy token, the legacy token
validator will
- add an audit annotation for the key-value pair

> `authentication.k8s.io/legacy-token-invalidated`:`secretName/Namepace`,

Reviewer: ```suggestion
  `authentication.k8s.io/legacy-token-invalidated`:`secretName/Namepace`,
```

Author: Revised.

- increment the `invalid_legacy_auto_token_uses_total` metric count,
- update the secret label `kubernetes.io/legacy-token-last-used` with the new date,
- return an error indicating that the legacy token has been invalidated.

> If received this validation error, users can update the secret to remove the

Reviewer: Please be careful when spelling "secret" or "Secret".
The former is a generic word and the second is treated as an object of a resource type.

Author: Revised.

`kubernetes.io/legacy-token-invalid-since` label to temporarily allow use of
this token.

Here's an example of an auto-generated legacy token that has been marked with the
`kubernetes.io/legacy-token-last-used` and `kubernetes.io/legacy-token-invalid-since`
labels:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: build-robot-secret
  namespace: default
  labels:
    kubernetes.io/legacy-token-last-used: 2022-10-24
    kubernetes.io/legacy-token-invalid-since: 2023-10-25
  annotations:
    kubernetes.io/service-account.name: build-robot
type: kubernetes.io/service-account-token
```

{{< note >}}
The difference between the manually created legacy token and
the auto-generated legacy token is that only auto-generated legacy token is
bi-directionally referenced by the ServiceAccount, like:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: build-robot
  namespace: default
secrets:
  - name: build-robot-secret
```

If you want to use the manually created legacy token, please do not add the

> secret name to the `Secrets` field of the service account.

Reviewer: Again, please check your whole PR for spelling ...

```suggestion
Secret name to the `secrets` field of the ServiceAccount.
```

Author: Checked and revised.

   admission controller adds `imagePullSecrets`, copying them from the `ServiceAccount`.

### Legacy ServiceAccount token tracking controller

{{< feature-state for_k8s_version="v1.28" state="stable" >}}

This controller generates a ConfigMap called
`kube-system/kube-apiserver-legacy-service-account-token-tracking` in the
`kube-system` namespace. The ConfigMap records the timestamp when legacy service
account tokens began to be monitored by the system.

### Legacy ServiceAccount token cleaner

{{< feature-state for_k8s_version="v1.29" state="beta" >}}

The legacy ServiceAccount token cleaner runs as part of the
`kube-controller-manager` and checks every 24 hours to see if any auto-generated
legacy ServiceAccount token has not been used in a *specified amount of time*.
If so, the cleaner marks those tokens as invalid.

The cleaner works by first checking the ConfigMap created by the control plane
(provided that `LegacyServiceAccountTokenTracking` is enabled). If the current
time is a *specified amount of time* after the date in the ConfigMap, the
cleaner then loops through the list of Secrets in the cluster and evaluates each

> Secret that has the type `kubernetes.io/service-account-token`.

Reviewer: These are implementation details, right?
As a user or an operator, skipping this paragraph doesn't hurt, right?

Reviewer: In other words, is there a reason for a user to care about the ConfigMap thing?

Author: Yeah. The cleaner initially inspects the tracking date within the ConfigMap to assess if the tracking period is sufficient. Without understanding the role of the ConfigMap, users might be puzzled as to why certain secrets that are seemingly old enough are being invalidated right away.

If a Secret meets all of the following conditions, the cleaner marks it as
invalid:

- The Secret is auto-generated, meaning that it is bi-directionally referenced
  by a ServiceAccount.
- The Secret is not currently mounted by any pods.
- The Secret has not been used in a *specified amount of time* since it was
  created or since it was last used.

The cleaner marks a Secret invalid by adding a label called
`kubernetes.io/legacy-token-invalid-since` to the Secret, with the current date
as the value. If an invalid Secret is not used in a *specified amount of time*,
the cleaner will delete it.

{{< note >}}

> All the *specified amount of time* above defaults to one year. Users can

Reviewer 4: ```suggestion
All the *specified amount of time* above defaults to one year. The cluster administrator can
```

If you launch a new Pod into the `examplens` namespace, it can use the `myserviceaccount`
service-account-token Secret that you just created.

{{< note >}}

> The way that Kubernetes is able to tell the difference between a manually
> created legacy token and an auto-generated legacy token is by assuming that
> only auto-generated legacy tokens are bi-directionally referenced by the
> ServiceAccount. For example:

Reviewer 4: ```suggestion
Kubernetes distinguishes auto-generated legacy tokens from manually created legacy tokens by
checking for a reference from the ServiceAccount's `secrets` field. For example:
```

Author: According to @zshihang's suggestion, I added one paragraph to explain the difference of auto-generated tokens and the manually generated tokens in the `Auto-generated legacy ServiceAccount token clean up` section. Thus I deleted the difference here. I also revised the "note" to the "warning" to acquire the user's attention. (I added the warning here as previously I saw the stack-overflow discussion about manually creating the legacy token. They think it's necessary to add the secret name to the service account 'secerts' field after creating the secret.)

Reviewer 3: Thanks, but please change the `warning` back to a `note`. Maybe a `caution`.

`warning` is reserved for “you must read this specific detail or your cluster will not be secure” levels of attention, eg “**Warning**: _do not copy the private key for the trust anchor and commit it to a public Git repo_”.

Author: Revised to 'caution'.

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: build-robot
  namespace: default
secrets:
  - name: build-robot-secret # triggers automatic legacy token generation
                             # (only on older clusters - automatic legacy token generation has been deprecated)
                             # usually NOT present for a manually generated token
```

> So please refrain from including manually created Secrets in the `secrets` field
> of a ServiceAccount.

Reviewer 4: ```suggestion
Do not reference manually created Secrets in the `secrets` field of a ServiceAccount.
```
